### ISSUE TITLE
新手求助

### ISSUE THREAD

**Issue by syghhhhh** [Moon0316/T2A#1]
https://github.com/Moon0316/T2A/blob/d0bb36ca554910110f38fe6170eea4a2800d85e2/dataloader.py#L65
这里我跑起来会出错，false少了一层维度，目的是生成类似[[-0.2113,  0.1069,  0.1805,  ..., -0.0475, -0.1105, -0.1938],[False]]这样为一行的数据吗。
hubert提取音频特征加入的这个False有什么用处吗，是不是可以不要这个直接做

**Comment by Moon0316** [comment]
可以检查一下arch是不是'chinese_hubert_large', 报错的地方这里跳转到的是提取wav2vec特征的代码，但是你提取的是hubert特征。正确提取的hubert特征是一个长度为24的列表，列表里每个元素是hubert的某一层的输出，它是一个tuple，我们只用tuple的第0维

**Comment by syghhhhh** [comment]
哦对 我用的是chinese_hubert_large，选错标的dataloader.py代码位置了。
python utils/generate_hubert.py --input_dir ./data/wavs --output_dir ./data/wav_features 
logits = model.extract_features(**inputs)
这里得到的logits形如
(tensor([[[-0.2113,  0.1069,  0.1805,  ..., -0.0475, -0.1105, -0.1938],
          [-0.2233,  0.3013,  0.1305,  ..., -0.0801, -0.1136, -0.0881],
          [-0.3120,  0.2703,  0.1377,  ..., -0.1692, -0.0800, -0.0114],
          ...,
          [-0.2075,  0.0674, -0.0291,  ..., -0.3489, -0.0023, -0.2074],
          [-0.2458,  0.0571,  0.0197,  ..., -0.3477,  0.0518, -0.3337],
          [-0.0831,  0.0114,  0.1610,  ...,  0.0386,  0.0269, -0.0251]]],
        dtype=torch.float16),
 tensor([[False, False, False, False, False, False, False, False, False, False,...False, False, False, False, False, False, False, False, False, False]]))
len(logits)==2  下一行这里logits[2]会出错

**Comment by Moon0316** [comment]
不好意思，这里是因为我们改了fairseq里hubert提取特征的源代码，因为原始的extract_features函数只返回最后一层输出，但是这里我们想要的是24层的输出，为此您可以把faiseq/fairseq/models/hubert/hubert.py 里的extract_features函数的返回值由```return feature, res["padding_mask"]```  改成 ```return feature, res["padding_mask"], res["layer_results"]```, 大概是在hubert.py文件的543行的位置。感谢您帮忙发现问题！

**Comment by syghhhhh** [comment]
非常不好意思，因为水平不足，请问可否上传一下改动之后的hubert.py。我试着修改了extract_features
` 
   def extract_features(
        self,
        source: torch.Tensor,
        padding_mask: Optional[torch.Tensor] = None,
        mask: bool = True,
        ret_conv: bool = False,
        output_layer: Optional[int] = None,
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        # output_layer=24
        res = self.forward(
            source,
            padding_mask=padding_mask,
            mask=mask,
            features_only=False,
            output_layer=output_layer,
        )
        feature = res["features"] if ret_conv else res["x"]
        # return feature, res["padding_mask"]
        return feature, res["padding_mask"], res["layer_results"]
`
但是没有什么进展，只加res["layer_results"]前面forward我不清楚怎么改感觉是解决不了问题，非常感谢您的指点。
ps：generate_mfcc.py 这里有笔误，应为`    d_mfcc_feat2 = python_speech_features.base.delta(input_values, 2)
    input_values = np.hstack((input_values, d_mfcc_feat, d_mfcc_feat2))`

**Comment by Moon0316** [comment]
您是对的，hubert.py确实不止一处修改，改过的hubert.py文件已经更新在仓库里了。再次感谢您的指正
